gateway/ln-gateway/src/gateway_module_ng/mod.rs

```diff
@@ -0,0 +1,287 @@
+            operation_id,
+            contract: contract.clone(),
+            state: SendSMState::Sending(Sending {
+                max_delay,
+                max_fee_msat,
+                invoice,
+                claim_keypair: self.keypair,
+            }),
+        });
+
+        self.client_ctx
+            .manual_operation_start(
+                operation_id,
+                LightningCommonInit::KIND.as_str(),
+                GatewayOperationMetaNG,
+                vec![self.client_ctx.make_dyn_state(send_sm)],
+            )
+            .await
+    }
+
+    pub async fn subscribe_send(
+        &self,
+        operation_id: OperationId,
+        contract: OutgoingContract,
+    ) -> Result<[u8; 32], Signature> {
+        let mut stream = self.notifier.subscribe(operation_id).await;
+
+        loop {
+            if let Some(GatewayClientStateMachinesNG::Send(state)) = stream.next().await {
+                match state.state {
+                    SendSMState::Sending(..) => {}
+                    SendSMState::Claiming(preimage, ..) => return Ok(preimage),
+                    SendSMState::Cancelled => {
+                        return Err(self.keypair.sign_schnorr(contract.forfeit_message()))
+                    }
+                }
+            }
+        }
+    }
+
+    pub async fn start_receive_state_machine(
+        &self,
+        operation_id: OperationId,
+        contract: IncomingContract,
+    ) -> anyhow::Result<()> {
+        let refund_keypair = self.keypair;
+
+        let client_output = ClientOutput::<LightningOutput, GatewayClientStateMachinesNG> {
+            output: LightningOutput::Incoming(contract.clone()),
+            state_machines: Arc::new(move |txid, out_idx| {
+                vec![GatewayClientStateMachinesNG::Receive(ReceiveStateMachine {
+                    operation_id,
+                    contract: contract.clone(),
+                    out_point: OutPoint { txid, out_idx },
+                    state: ReceiveSMState::Funding(refund_keypair),
+                })]
+            }),
+        };
+
+        let client_output = self.client_ctx.make_client_output(client_output);
+        let transaction = TransactionBuilder::new().with_output(client_output);
+
+        self.client_ctx
+            .finalize_and_submit_transaction(
+                operation_id,
+                LightningCommonInit::KIND.as_str(),
+                |_, _| GatewayOperationMetaNG,
+                transaction,
+            )
+            .await?;
+
+        Ok(())
+    }
+
+    pub async fn subscribe_receive(&self, operation_id: OperationId) -> Option<[u8; 32]> {
+        let mut stream = self.notifier.subscribe(operation_id).await;
+
+        loop {
+            if let Some(GatewayClientStateMachinesNG::Receive(state)) = stream.next().await {
+                match state.state {
+                    ReceiveSMState::Funding(..) => {}
+                    ReceiveSMState::Rejected(..) => return None,
```

Comment:
This function should probably return the `UpdateStreamOrOutcome` similar to how we do it in legacy LN rather than just a preimage, so that we know what went wrong (rejected vs failure, etc)